File: py_tf/py_tf/tf_listener.py
import rclpy
from rclpy.node import Node
import rclpy.time
from tf2_ros import TransformListener, Buffer
from geometry_msgs.msg import TransformStamped
from tf_transformations import euler_from_quaternion


class TfListener(Node):
    def __init__(self, name):
        super().__init__(name)
        self.buffer = Buffer()
        self.tf_listener = TransformListener(node=self, buffer=self.buffer)
        self.timer = self.create_timer(1, self.get_transform)

    def get_transform(self):
        """获取坐标变换"""
        try:
            result = self.buffer.lookup_transform('base_link', 'bottle_link',
                # 获取transform的时间段，0代表获取最新的
                rclpy.time.Time(seconds=0.0), 
                # 超时时间
                rclpy.time.Duration(seconds=1.0))
            # 解析结果
            transform = result.transform
            # 四元数转欧拉角
            euler = euler_from_quaternion([
                transform.rotation.x,
                transform.rotation.y,
                transform.rotation.z,
                transform.rotation.w,
            ])
            self.get_logger().info(f"平移:{transform.translation}\n旋转四元数:{transform.rotation}\n旋转欧拉角:{euler}")
        except Exception as e:
            self.get_logger().warn(f"不能够获取坐标变换，原因: {str(e)}")


def main():
    rclpy.init()
    node = TfListener("tf_listener")
    rclpy.spin(node)
    rclpy.shutdown()

File: cpp_tf/src/tf_listener.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2/LinearMath/Quaternion.hpp"
#include "tf2/utils.h"                  // 提供 tf2::getEulerYPR 函数
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp" 

using namespace std::chrono_literals;
using TransformListener = tf2_ros::TransformListener;
using TransformStamped = geometry_msgs::msg::TransformStamped;

class DynamicTfBroadcaster : public rclcpp::Node
{
public:
DynamicTfBroadcaster(std::string name) : Node(name)
    {
        buffer_  = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<TransformListener>(*buffer_, this);
        timer_ = this->create_wall_timer(1s, std::bind(&DynamicTfBroadcaster::get_transform, this));
    }

private:
    void get_transform(void)
    {   
        try {
            auto result = buffer_->lookupTransform("base_link", "target_point",
                rclcpp::Time(0.0f), rclcpp::Duration::from_seconds(1.0f));
    
            RCLCPP_INFO(this->get_logger(), "平移 %f %f %f", \
                result.transform.translation.x,
                result.transform.translation.y,
                result.transform.translation.z
            );
            RCLCPP_INFO(this->get_logger(), "旋转(四元数) %f %f %f %f", \
                result.transform.rotation.x,
                result.transform.rotation.y,
                result.transform.rotation.z,
                result.transform.rotation.w
            );
            double roll, pitch, yaw;
            const auto &q = result.transform.rotation;
            tf2::getEulerYPR(q, yaw, pitch, roll);
            RCLCPP_INFO(this->get_logger(), "旋转(欧拉角) %f %f %f", \
                roll,
                pitch,
                yaw
            );
        } catch(tf2::TransformException &e) {
            RCLCPP_WARN(this->get_logger(), "异常: %s", e.what());
        }
        
    }
    
private:
    std::shared_ptr<TransformListener> tf_listener_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::unique_ptr<tf2_ros::Buffer> buffer_;
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DynamicTfBroadcaster>("tf_listener");
    rclcpp::spin(node);
    rclcpp::shutdown();
}